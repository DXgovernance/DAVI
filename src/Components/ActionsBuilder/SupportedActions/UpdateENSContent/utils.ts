import contenthash from 'content-hash';
import { MAINNET_ID, LOCALHOST_ID, GOERLI_ID } from 'utils';
import { utils } from 'ethers';

const DEFAULT_NAMEHASH =
  '0x0000000000000000000000000000000000000000000000000000000000000000';

export const convertToNameHash = (name: string) => {
  if (!name) {
    return DEFAULT_NAMEHASH;
  }
  return utils.namehash(name);
};

export const convertToContentHash = (ipfsHash: string) => {
  if (!ipfsHash) {
    return '';
  }
  return contenthash.fromIpfs(ipfsHash);
};

export const convertToIpfsHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const getIpfsUrl = (ipfsHash: string) => {
  const ipfsRoot = 'ipfs://';
  return `${ipfsRoot}${ipfsHash}`;
};

export const isValidChainId = (chainId: number) => {
  if (!chainId || chainId === LOCALHOST_ID) {
    return MAINNET_ID;
  } else {
    return chainId;
  }
};

export const isAvailableOnENS = (chainId: number) => {
  const validChainId = isValidChainId(chainId);
  const ensNetworks = [MAINNET_ID, GOERLI_ID];
  return ensNetworks.includes(validChainId);
};
