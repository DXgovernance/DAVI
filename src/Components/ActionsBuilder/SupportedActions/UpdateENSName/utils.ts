import namehash from 'eth-ens-namehash';
import contenthash from 'content-hash';
import { useState, useEffect } from 'react';

export const convertToNameHash = (name: string) => namehash.hash(name);

export const convertToContentHash = (ipfsHash: string) =>
  contenthash.fromIpfs(ipfsHash);

export const convertToIPFSHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const useDebounce = (value: string, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
};
