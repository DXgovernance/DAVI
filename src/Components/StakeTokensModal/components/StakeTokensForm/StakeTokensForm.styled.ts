import styled, { css } from 'styled-components';
import { Heading } from 'Components/Primitives/Typography';
import TokenAmountInput from 'Components/Primitives/Forms/TokenAmountInput';
import { Button } from 'Components/Primitives/Button';

export const GuestContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  color: black;
  @media only screen and (min-width: 769px) {
    padding: 20px 40px;
  }
`;

export const DaoBrand = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;
export const DaoIcon = styled.img`
  height: 3rem;
  width: 3rem;
`;

export const DaoTitle = styled(Heading)`
  font-weight: ${({ theme }) => theme.fontWeights.medium};
  color: ${({ theme }) => theme.colors.text};
`;

export const InfoItem = styled.div`
  display: flex;
  font-size: ${({ theme }) => theme.fontSizes.body};
  color: ${({ theme }) => theme.colors.card.grey};
  margin-bottom: 0.4rem;
`;

export const BalanceWidget = styled.div`
  width: 100%;
  padding: 10px;
  display: flex;
  flex-direction: column;
  margin: 1rem 0;
  border: 1px solid ${({ theme }) => theme.colors.muted};
  border-radius: ${({ theme }) => theme.radii.curved};
`;

export const InfoRow = styled.div`
  display: flex;
  width: 100%;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin-top: 10px;
`;

export const BaseFont = css`
  font-size: 12px;

  @media only screen and (min-width: 769px) {
    font-size: 14px;
  }
`;

export const InfoLabel = styled.span`
  ${BaseFont}
  color: ${({ theme }) => theme.colors.card.grey};
`;

export const InfoValue = styled.span`
  ${BaseFont}
  font-weight: bold;
  flex-wrap: wrap;
  display: inline-flex;
  align-items: center;
  color: ${({ theme }) => theme.colors.text};
`;

export const InfoOldValue = styled(InfoValue)`
  color: ${({ theme }) => theme.colors.muted};
  display: inline-flex;
  align-items: center;
`;

export const StakeAmountInput = styled(TokenAmountInput)`
  font-size: 20px;
  font-weight: 600;
  border: none;
  outline: none;
  width: inherit;
  font-family: inherit;
`;

export const ActionButton = styled(Button)`
  width: 100%;
  margin-top: 22px;
  self-align: flex-end;
`;
