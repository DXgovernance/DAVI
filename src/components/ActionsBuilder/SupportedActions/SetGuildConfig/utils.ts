import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
