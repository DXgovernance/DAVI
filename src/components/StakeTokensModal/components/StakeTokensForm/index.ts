export { default as StakeTokensForm } from './StakeTokensForm';
