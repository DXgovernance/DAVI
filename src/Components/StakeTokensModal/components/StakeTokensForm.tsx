import dxIcon from '../../../assets/images/dxdao-icon.svg';
import { Loading } from '../../Primitives/Loading';
import { formatUnits } from 'ethers/lib/utils';
import moment from 'moment';
import { useMemo, useState } from 'react';
import { FiArrowRight, FiInfo } from 'react-icons/fi';
import { useHistory, useLocation } from 'react-router-dom';
import { MAX_UINT } from 'utils';
import {
  GuestContainer,
  DaoBrand,
  DaoIcon,
  DaoTitle,
  InfoItem,
  BalanceWidget,
  InfoRow,
  InfoLabel,
  InfoValue,
  InfoOldValue,
  StakeAmountInput,
  ActionButton,
} from './StakeTokensForm.styled';
import { Button } from 'old-components/Guilds/common/Button';
import useBigNumberToNumber from '../../../hooks/Guilds/conversions/useBigNumberToNumber';
import { BigNumber } from 'ethers';
import useVotingPowerPercent from '../../../hooks/Guilds/guild/useVotingPowerPercent';
import { StakeTokensFormsProps } from '../types';

export const StakeTokensForm = ({
  token,
  userVotingPower,
  createTransaction,
  guild,
  isRepGuild,
}: StakeTokensFormsProps) => {
  const [stakeAmount, setStakeAmount] = useState<BigNumber>(null);

  const isStakeAmountValid = useMemo(
    () =>
      stakeAmount?.gt(0) &&
      token?.info?.decimals &&
      stakeAmount.lte(token?.balance),
    [stakeAmount, token?.balance, token?.info]
  );

  const lockTokens = async () => {
    if (!isStakeAmountValid) return;

    createTransaction(
      `Lock ${formatUnits(stakeAmount, token?.info?.decimals)} ${
        token?.info?.symbol
      } tokens`,
      async () => guild?.contract?.lockTokens(stakeAmount)
    );
  };

  const approveTokenSpending = async () => {
    if (!isStakeAmountValid) return;

    createTransaction(
      `Approve ${token?.info?.symbol} token spending`,
      async () => token?.contract.approve(guild?.config?.tokenVault, MAX_UINT)
    );
  };
  const roundedBalance = useBigNumberToNumber(
    token?.balance,
    token?.info?.decimals,
    4
  );

  const currentvotingPowerPercent = useVotingPowerPercent(
    userVotingPower,
    guild?.config?.totalLocked,
    3
  );
  const nextVotingPowerPercent = useVotingPowerPercent(
    stakeAmount?.add(userVotingPower),
    stakeAmount?.add(guild?.config?.totalLocked),
    3
  );

  const history = useHistory();
  const location = useLocation();
  return (
    <GuestContainer data-testid="stake-tokens-modal">
      <DaoBrand>
        <DaoIcon src={dxIcon} alt={'DXdao Logo'} />
        <DaoTitle>
          {guild?.config?.name || (
            <Loading text loading skeletonProps={{ width: 100 }} />
          )}
        </DaoTitle>
      </DaoBrand>
      {!isRepGuild && (
        <InfoItem>
          {guild?.config?.lockTime ? (
            `${moment
              .duration(guild?.config?.lockTime.toNumber(), 'seconds')
              .humanize()} staking period`
          ) : (
            <Loading loading text skeletonProps={{ width: 200 }} />
          )}{' '}
        </InfoItem>
      )}

      {!isRepGuild && (
        <BalanceWidget>
          <InfoRow>
            <InfoLabel>Balance:</InfoLabel>
            <InfoValue>
              {token?.balance && token?.info ? (
                roundedBalance
              ) : (
                <Loading loading text skeletonProps={{ width: 30 }} />
              )}{' '}
              {token?.info?.symbol || (
                <Loading loading text skeletonProps={{ width: 10 }} />
              )}
            </InfoValue>
          </InfoRow>
          <InfoRow>
            <StakeAmountInput
              value={stakeAmount}
              onChange={setStakeAmount}
              decimals={token?.info?.decimals}
              data-testid="stake-amount-input"
            />
            <Button
              data-testid="stake-amount-max-button"
              onClick={() => setStakeAmount(token?.balance)}
            >
              Max
            </Button>
          </InfoRow>
        </BalanceWidget>
      )}
      {isRepGuild && (
        <InfoRow>
          <InfoLabel>Balance</InfoLabel>
          <InfoValue>
            {token?.balance && token?.info ? (
              roundedBalance
            ) : (
              <Loading loading text skeletonProps={{ width: 30 }} />
            )}{' '}
            {token?.info?.symbol || (
              <Loading loading text skeletonProps={{ width: 10 }} />
            )}
          </InfoValue>
        </InfoRow>
      )}
      <InfoRow>
        <InfoLabel>Your voting power</InfoLabel>
        <InfoValue>
          {isStakeAmountValid ? (
            <>
              <InfoOldValue>
                {currentvotingPowerPercent != null ? (
                  `${currentvotingPowerPercent}%`
                ) : (
                  <Loading loading text skeletonProps={{ width: 40 }} />
                )}{' '}
                <FiArrowRight />
              </InfoOldValue>{' '}
              <strong>
                {nextVotingPowerPercent != null ? (
                  `${nextVotingPowerPercent}%`
                ) : (
                  <Loading loading text skeletonProps={{ width: 40 }} />
                )}
              </strong>
            </>
          ) : (
            '-'
          )}
        </InfoValue>
      </InfoRow>
      {!isRepGuild && (
        <InfoRow>
          <InfoLabel>Unlock Date</InfoLabel>
          <InfoValue>
            {isStakeAmountValid ? (
              <>
                <strong>
                  {moment()
                    .add(guild?.config?.lockTime.toNumber(), 'seconds')
                    .format('MMM Do, YYYY - h:mm a')}
                </strong>{' '}
                <FiInfo />
              </>
            ) : (
              '-'
            )}
          </InfoValue>
        </InfoRow>
      )}
      {!isRepGuild ? (
        stakeAmount && token?.allowance?.gte(stakeAmount) ? (
          <ActionButton
            data-testid="lock-token-spending"
            disabled={!isStakeAmountValid}
            onClick={lockTokens}
          >
            Lock{' '}
            {token?.info?.symbol || (
              <Loading loading text skeletonProps={{ width: 10 }} />
            )}
          </ActionButton>
        ) : (
          <ActionButton
            disabled={!isStakeAmountValid}
            onClick={approveTokenSpending}
            data-testid="approve-token-spending"
          >
            Approve{' '}
            {token?.info?.symbol || (
              <Loading loading text skeletonProps={{ width: 10 }} />
            )}{' '}
            Spending
          </ActionButton>
        )
      ) : (
        <ActionButton
          onClick={() => history.push(location.pathname + '/proposalType')}
        >
          Mint Rep
        </ActionButton>
      )}
    </GuestContainer>
  );
};
