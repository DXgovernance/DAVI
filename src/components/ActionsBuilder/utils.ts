import { DecodedAction } from './types';
import { Permission } from './types';
import Web3 from 'web3';
import { NULL_SIGNATURE } from 'utils';
import { SupportedAction } from './types';

const web3 = new Web3();

export const getPermissionArgs = (
  decodedActions: DecodedAction[]
): Permission[] => {
  const permissions = decodedActions.map(decodedAction => {
    // Native transfer of ETH does not have a function
    if (
      decodedAction?.decodedCall?.callType ===
        SupportedAction.NATIVE_TRANSFER ||
      decodedAction?.decodedCall?.callType === SupportedAction.RAW_TRANSACTION
    ) {
      return {
        from: decodedAction?.decodedCall?.from,
        to: decodedAction?.decodedCall?.to,
        callType: decodedAction?.decodedCall?.callType,
        functionSignature: NULL_SIGNATURE,
      };
    }
    const functionInputs = decodedAction?.decodedCall?.function?.inputs?.map(
      input => input.type
    );
    const functionName = decodedAction?.decodedCall?.function?.name;
    const functionSignature = web3.eth.abi.encodeFunctionSignature(
      `${functionName}(${functionInputs.join(',')})`
    );
    return {
      from: decodedAction?.decodedCall?.from,
      to: decodedAction?.decodedCall?.to,
      callType: decodedAction?.decodedCall?.callType,
      functionSignature,
    };
  });
  return permissions;
};
