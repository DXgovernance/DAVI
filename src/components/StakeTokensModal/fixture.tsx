import { BigNumber } from 'ethers';
import { GuildConfigProps } from 'hooks/Guilds/ether-swr/guild/useGuildConfig';
import { ERC20 } from 'types/contracts';
import { ERC20Guild } from 'types/contracts/ERC20Guild';
import StakeTokensForm from './components/StakeTokensForm/StakeTokensForm';

export const mockStakeTokensFormProps = {
  token: {
    name: 'Mock Token',
    allowance: BigNumber.from(1000000),
    balance: BigNumber.from(1000000),
    info: {
      decimals: 18,
      symbol: 'REP',
      name: 'mockName',
      totalSupply: BigNumber.from(0),
    },
    contract: {
      approve: () => {},
    } as unknown as ERC20,
  },
  userVotingPower: BigNumber.from(0),
  guild: {
    contract: {} as ERC20Guild,
    config: {
      name: 'mockGuildName',
      totalLocked: BigNumber.from(0),
      tokenVault: 'mockTokenVault',
      lockTime: BigNumber.from(0),
    } as unknown as GuildConfigProps,
    totalLocked: BigNumber.from(0),
  },
  createTransaction: () => {},
  isRepGuild: false,
};

export const mockStakeTokensModalProps = {
  isOpen: true,
  onDismiss: () => {},
  StakeTokensForm: StakeTokensForm,
  StakeTokensFormProps: mockStakeTokensFormProps,
};
