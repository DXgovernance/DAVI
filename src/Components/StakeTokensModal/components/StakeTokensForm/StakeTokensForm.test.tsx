import StakeTokensForm from './StakeTokensForm';
import { render } from 'utils/tests';
import { mockStakeTokensFormProps } from '../../fixture';

describe('StakeTokensForm', () => {
  it('StakeTokensForm renders properly', () => {
    const { container } = render(
      <StakeTokensForm {...mockStakeTokensFormProps} />
    );
    expect(container).toMatchSnapshot();
  });
});
