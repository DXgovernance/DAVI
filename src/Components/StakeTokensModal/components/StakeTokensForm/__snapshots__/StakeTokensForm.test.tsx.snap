// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakeTokensForm StakeTokensForm renders properly 1`] = `
<div>
   
  <div
    class="sc-pGacB fMzOOg"
    data-testid="stake-tokens-modal"
  >
    <div
      class="sc-jrAFXE iOZqsn"
    >
      <img
        alt="DXdao Logo"
        class="sc-kEjbQP gWbxpV"
        src="dxdao-icon.svg"
      />
      <h2
        class="sc-dlfnuX sc-iqHYmW bvZvxV dbvXAA"
      >
        mockGuildName
      </h2>
    </div>
    <div
      class="sc-crrszt cGMjRh"
    >
      a few seconds staking period
    </div>
    <div
      class="sc-dQoVA bHndUh"
    >
      <div
        class="sc-bqyKOL bFgWEW"
      >
        <span
          class="sc-kstqJO eCKOSo"
        >
          balance
          :
        </span>
        <span
          class="sc-hBEYId jtmCId"
        >
          0
           
          REP
        </span>
      </div>
      <div
        class="sc-bqyKOL bFgWEW"
      >
        <div
          class="sc-eCstlR cTttQe"
        >
          <div
            class="sc-jSgvzq vMswc"
          />
          <input
            autocomplete="off"
            autocorrect="off"
            class="sc-hKgJUU czqcGJ sc-fFucqa cMBanK"
            data-testid="stake-amount-input"
            inputmode="decimal"
            maxlength="79"
            minlength="1"
            pattern="^[0-9]*[.,]?[0-9]*$"
            placeholder="0.0"
            spellcheck="false"
            type="text"
            value="0.0"
          />
          <div
            class="sc-jSgvzq iDoZOX"
          />
        </div>
        <button
          class="sc-gKseQn dObHhs"
          data-testid="stake-amount-max-button"
        >
          max
        </button>
      </div>
    </div>
    <div
      class="sc-bqyKOL bFgWEW"
    >
      <span
        class="sc-kstqJO eCKOSo"
      >
        balance
      </span>
      <span
        class="sc-hBEYId jtmCId"
      >
        0
         
        REP
      </span>
    </div>
    <div
      class="sc-bqyKOL bFgWEW"
    >
      <span
        class="sc-kstqJO eCKOSo"
      >
        yourVotingPower
      </span>
      <span
        class="sc-hBEYId jtmCId"
      >
        -
      </span>
    </div>
    <div
      class="sc-bqyKOL bFgWEW"
    >
      <span
        class="sc-kstqJO eCKOSo"
      >
        unlockDate
      </span>
      <span
        class="sc-hBEYId jtmCId"
      >
        -
      </span>
    </div>
    <button
      class="sc-gKseQn sc-bkzYnD dObHhs brvGTX"
      data-testid="lock-token-spending"
      disabled=""
    >
      Lock
       
      REP
    </button>
  </div>
   
</div>
`;
